test/suites-ostree/staged-failed-upgrade.robot

```diff
@@ -0,0 +1,82 @@
+*** Settings ***
+Documentation       Tests related to upgrading MicroShift
+
+Resource            ../resources/common.resource
+Resource            ../resources/ostree.resource
+Library             Collections
+
+Suite Setup         Setup
+Suite Teardown      Teardown
+
+Test Tags           ostree
+
+
+*** Variables ***
+${USHIFT_HOST}      ${EMPTY}
+${USHIFT_USER}      ${EMPTY}
+
+${TARGET_REF}       ${EMPTY}
+
+
+*** Test Cases ***
+First Failed Staged Upgrade Should Restore Backup
+    [Documentation]    Performs an upgrade, when first unhealthy a restore is performend
+
+    Wait For Healthy System
+
+    ${initial_backup}=    Get Future Backup Name For Current Boot
+
+    TestAgent.Add Action For Next Deployment    1    fail_greenboot
+
+    # We don't expect a rollback here since after the first failure
+    # things should continue as desired. This test validates the
+    # restore function is working when the first boot is unhealthy.
+    Deploy Commit Not Expecting A Rollback    ${TARGET_REF}    ${TRUE}
+
+    Wait For Healthy System
+
+    Validate Backup Is Restored    ${initial_backup}
+    Expected Boot Count    3
+
+
+*** Keywords ***
+Setup
+    [Documentation]    Test suite setup
+    Check Required Env Variables
+    Should Not Be Empty    ${TARGET_REF}    TARGET_REF variable is required
+    Login MicroShift Host
+    Setup Kubeconfig
+
+Teardown
+    [Documentation]    Test suite teardown
+    Logout MicroShift Host
+
+Validate Backup Is Restored
+    [Documentation]    Validate that the desired backup exists and is the only one that exists
+    [Arguments]    ${backup_name}
+
+    Backup Should Exist    ${backup_name}
+
+    ${stdout}    ${rc}=    Execute Command
+    ...    cd ${BACKUP_STORAGE} && ls -d1 rhel-*
+    ...    sudo=False    return_rc=True
+
+    ${grep_query}=    Catenate    SEPARATOR=
+    ...    "Restoring backup for a rollback deployment to perform migration and try starting again"
```

Comment:
Looking for this message tells us that MicroShift thinks it is restoring the content, but it doesn't tell us that the content was actually restored. What can we do to prove it's the same data?

Could we put a marker file in the data directory, for example?

Reply:
> Looking for this message tells us that MicroShift thinks it is restoring the content, but it doesn't tell us that the content was actually restored. What can we do to prove it's the same data?

In this test the way we try and validate that, is by checking what MicroShift logs has happened and what the state of the backup directory is after. We record the initial backup, and make sure that is the only backup that exists in the filesystem, plus the log statement from MicroShift, we can be sure that things progressed as intended. If there was a failure in the restore MicroShift wouldn't start and that would break this test.

> Could we put a marker file in the data directory, for example?

I'm not opposed to this, however, in the backup/restore logic we put a lot of effort to be self healing. Presenting a file marker might be too much of an implementation detail where it's usefulness might be limited to just this test. Maybe a consistent log statement would suffice for this? (ex: `step: prerun, backup: created | restored...`)

Reply:
I think the marker Doug is mentioning is created in test, TAL:
- https://github.com/openshift/microshift/blob/main/test/suites-ostree/backup-restore.robot#L41 
- https://github.com/openshift/microshift/blob/main/test/suites-ostree/rollback.robot#L35-L38

We could do that here, we just need to get the timing right, but that should be hard, there's plenty of time: after rebooting to newly staged deployment, shortly after starting MicroShift should create a backup, that's our window until the auto-reboot

Reply:
Ah see, thats way easier. Let me update the test to use that then. Do we even need to worry about the timing if we do what the restore does and just manually create a backup and create the marker there and kick off the deployment?

Reply:
I think that creating backup manually will cause MicroShift to fail, because it's not expecting it to exist (even for the same deployment, each boot produces different backup based on previous boot ID, we just keep the most recent one)